src/main/scala/com/cognite/sdk/scala/v1/Client.scala

```diff
@@ -38,72 +41,50 @@ final case class RequestSession[F[_]: Monad](
-        md.contentLength.forall(_ > 0) && md.contentType.exists(
-          // This is a bit ugly, but we're highly unlikely to use any other ContentType
-          // any time soon.
-          ct => ct.startsWith(MediaTypes.Json) || ct.startsWith("application/protobuf")
-        )
-      }
+//      .parseResponseIfMetadata { md =>
+//        md.contentLength.forall(_ > 0) && md.contentType.exists(
+//          // This is a bit ugly, but we're highly unlikely to use any other ContentType
+//          // any time soon.
+//          ct => ct.startsWith(MediaTypes.Json) || ct.startsWith("application/protobuf")
+//        )
+//      }
     clientTag match {
       case Some(tag) => baseRequest.header("x-cdp-clienttag", tag)
       case None => baseRequest
     }
   }
 
   private def parseResponse[T, R](uri: Uri, mapResult: T => R)(
-      implicit decoder: Decoder[Either[CdpApiError, T]]
+      implicit decoder: Decoder[T]
   ) =
-    asJson[Either[CdpApiError, T]].mapWithMetadata((response, metadata) =>
+    asJsonEither[CdpApiError, T].mapWithMetadata((response, metadata) =>
       response match {
-        case Left(value) =>
-          // As of 2020-02-17, content length is not reliably set.
-          // Checking the content type explicitly for JSON and protobuf isn't
-          // ideal either, but we're not likely to add other content types
-          // any time soon.
-          val shouldHaveParsed = // metadata.contentLength.exists(_ > 0) &&
-            metadata.contentType.exists(_.startsWith(MediaTypes.Json)) ||
-              metadata.contentType.exists(_.startsWith("application/protobuf"))
-          if (shouldHaveParsed) {
-            throw SdkException(
-              s"Failed to parse response, reason: ${value.error.getMessage}",
-              Some(uri),
-              metadata.header("x-request-id"),
-              Some(metadata.code)
-            )
-          } else {
-            val message = if (metadata.statusText.isEmpty) {
-              "Unknown error (no status text)"
-            } else {
-              metadata.statusText
-            }
-            throw SdkException(
-              message,
-              Some(uri),
-              metadata.header("x-request-id"),
-              Some(metadata.code)
-            )
-          }
-        case Right(Left(cdpApiError)) =>
+        case Left(DeserializationException(_, error)) =>
+          throw SdkException(
+            s"Failed to parse response, reason: ${error.getMessage}",
+            Some(uri),
+            metadata.header("x-request-id"),
+            Some(metadata.code.code)
+          )
+        case Left(HttpError(cdpApiError, _)) =>
           throw cdpApiError.asException(uri"$uri", metadata.header("x-request-id"))
-        case Right(Right(value)) => mapResult(value)
+        case Right(value) => mapResult(value)
       }
     )
 
   private def unsafeBody[R](uri: Uri, response: Response[R]): R =
-    try response.unsafeBody
+    try response.body
```

Comment:
Yay! Is the comment below correct still or can we get rid of it?

Reply:
We should be able to get rid of all of unsafeBody, actually.